pkg/pillar/cmd/zedrouter/zedrouter.go

```diff
@@ -1239,59 +1256,92 @@ func findBridge(bridgeName string) (*netlink.Bridge, error) {
+		return "", nil
+	}
+	log.Functionf("getUlAddrs(%d) for %s/%s\n", ifnum,
+		networkID.String(), appID.String())
 
-	// Allocate bridgeIPAddr based on BridgeMac
-	log.Functionf("getUlAddrs(%d/%d for %s) bridgeMac %s\n",
-		ifnum, appNum, netInstStatus.UUID.String(),
-		status.BridgeMac.String())
-	var err error
-	var addr string
-	addr, err = lookupOrAllocateIPv4(ctx, netInstStatus,
-		status.BridgeMac)
+	// XXX or change type of VifInfo.Mac to avoid parsing?
+	mac, err = net.ParseMAC(ulStatus.Mac)
 	if err != nil {
-		log.Errorf("getUlAddrs: Bridge IP address allocation failed %s\n", err)
-		return bridgeIPAddr, appIPAddr, err
+		errStr := fmt.Sprintf("parse Mac fail: %v\n", err)
+		log.Errorf("getUlAddrs(%s, %s): fail: %s\n",
+			networkID.String(), appID.String(), err)
+		return "", errors.New(errStr)
+	}
+
+	ipAddr := ""
+	// for static IP Address
+	if ulStatus.AppIPAddr != nil {
+		ipAddr = ulStatus.AppIPAddr.String()
+		// the IP Address, should not be in dhcpRange
+		if netInstStatus.DhcpRange.Contains(ulStatus.AppIPAddr) {
+			errStr := fmt.Sprintf("static IP(%s) is in DhcpRange(%s, %s)",
+				ipAddr, netInstStatus.DhcpRange.Start.String(),
+				netInstStatus.DhcpRange.End.String())
+			log.Errorf("getUlAddrs(%s, %s): fail: %s",
+				networkID.String(), appID.String(), errStr)
+			return "", errors.New(errStr)
+		}
+		// IP Address must be inside the subnet range
+		if !netInstStatus.Subnet.Contains(ulStatus.AppIPAddr) {
+			errStr := fmt.Sprintf("out of subnet range(%s)", ipAddr)
+			log.Errorf("getUlAddrs(%s, %s): fail: %s",
+				networkID.String(), appID.String(), errStr)
+			return "", errors.New(errStr)
+		}
 	} else {
-		bridgeIPAddr = addr
-	}
-
-	if status.AppIPAddr != nil {
-		// Static IP assignment case.
-		// Note that appIPAddr can be in a different subnet.
-		// Assumption is that the config specifies a gateway/router
-		// in the same subnet as the static address.
-		appIPAddr = status.AppIPAddr.String()
-		recordIPAssignment(ctx, netInstStatus, status.AppIPAddr,
-			status.Mac)
-	} else if status.Mac != "" {
-		// XXX or change type of VifInfo.Mac to avoid parsing?
-		var mac net.HardwareAddr
-		mac, err = net.ParseMAC(status.Mac)
+		// get the app number for the underlay network entry
+		appNum, err := appNumOnUNetGet(ctx, networkID, appID)
 		if err != nil {
-			log.Fatal("ParseMAC failed: ", status.Mac, err)
+			errStr := fmt.Sprintf("App Number get failed: %v", err)
+			log.Errorf("getUlAddrs(%s, %s): fail: %s",
+				networkID.String(), appID.String(), errStr)
+			return "", errors.New(errStr)
+		}
+		hostCount := netInstStatus.DhcpRange.HostCount()
+		if appNum > hostCount {
+			errStr := fmt.Sprintf("IP Addr get fail: %d is outside DhcpRange",
+				appNum)
+			log.Errorf("getUlAddrs(%s%s): fail:%s",
+				networkID.String(), appID.String(), errStr)
+			log.Errorf("getUlAddrs(%s, %s): fail: %s",
+				networkID.String(), appID.String(), errStr)
+			return "", errors.New(errStr)
 		}
```

Comment:
This is duplicate of what lookupOrAllocateIPv4 does for the subnet check, right? So it can be removed.